model/task_lifecycle.go

```diff
@@ -727,21 +683,47 @@ func RestartItemsAfterVersion(cq *commitqueue.CommitQueue, project, version, cal
 	return catcher.Resolve()
 }
 
-func tryDequeueAndAbortCommitQueueVersion(t *task.Task, cq commitqueue.CommitQueue, caller string) error {
-	p, err := patch.FindOne(patch.ByVersion(t.Version))
+// DequeueAndRestartForTask restarts all items after the given task's version, aborts/dequeues the current version,
+// and sends an updated status to GitHub.
+func DequeueAndRestartForTask(cq *commitqueue.CommitQueue, t *task.Task, githubState message.GithubState, caller, reason string) error {
+	if cq == nil {
+		var err error
+		cq, err = commitqueue.FindOneId(t.Project)
+		if err != nil {
+			return errors.Wrapf(err, "can't get commit queue for id '%s'", t.Project)
+		}
+		if cq == nil {
+			return errors.Errorf("no commit queue found for '%s'", t.Project)
+		}
+	}
+	// this must be done before dequeuing so that we know which entries to restart
+	if err := RestartItemsAfterVersion(cq, t.Project, t.Version, caller); err != nil {
+		return errors.Wrap(err, "unable to restart versions")
+	}
+
+	p, err := patch.FindOneId(t.Version)
 	if err != nil {
-		return errors.Wrapf(err, "error finding patch")
+		return errors.Wrap(err, "unable to find patch")
 	}
 	if p == nil {
-		return errors.Errorf("No patch for task")
+		return errors.New("patch not found")
 	}
-
-	if !p.IsCommitQueuePatch() {
-		return nil
+	if err := tryDequeueAndAbortCommitQueueVersion(p, *cq, t.Id, caller); err != nil {
+		return err
```

Comment:
How should we handle the situation if this fails after we've already successfully restarted CQ items after the current version? Maybe we should add a log explicitly stating that we tried to dequeue and abort the current version but it is still going to run (and fail).

Reply:
This would only happen in the event that there's a database error and isn't a new behavior, and typically returning an error assumes it will get propagated to something that will log it (either to the user or to splunk at the tope level) so I'm not super concerned with adding another log here.